include/pybind11/pybind11.h

```diff
@@ -946,55 +944,56 @@ class module_ : public object {
     }
 
-    // Adds an object to the module using the given name.  Throws if an object with the given name
-    // already exists.
-    //
-    // overwrite should almost always be false: attempting to overwrite objects that pybind11 has
-    // established will, in most cases, break things.
+    /** \rst
+        Adds an object to the module using the given name.  Throws if an object with the given name
+        already exists.
+
+        ``overwrite`` should almost always be false: attempting to overwrite objects that pybind11 has
+        established will, in most cases, break things.
+    \endrst */
     PYBIND11_NOINLINE void add_object(const char *name, handle obj, bool overwrite = false) {
         if (!overwrite && hasattr(*this, name))
             pybind11_fail("Error during initialization: multiple incompatible definitions with name \"" +
                     std::string(name) + "\"");
 
         PyModule_AddObject(ptr(), name, obj.inc_ref().ptr() /* steals a reference */);
     }
-
-private:
-#if PY_MAJOR_VERSION >= 3
-    friend module_ detail::create_top_level_module(const char *, const char *, PyModuleDef *);
-
-    explicit module_(const char *name, const char *doc, PyModuleDef *def) {
-        def = new (def) PyModuleDef {  // Placement new (not an allocation).
-            /* m_base */     PyModuleDef_HEAD_INIT,
-            /* m_name */     name,
-            /* m_doc */      options::show_user_defined_docstrings() ? doc : nullptr,
-            /* m_size */     -1,
-            /* m_methods */  nullptr,
-            /* m_slots */    nullptr,
-            /* m_traverse */ nullptr,
-            /* m_clear */    nullptr,
-            /* m_free */     nullptr
-        };
-        m_ptr = PyModule_Create(def);
-        if (m_ptr == nullptr)
-            pybind11_fail("Internal error in module_::module_()");
-        inc_ref();
-    }
-#endif
 };
 
 // When inside a namespace (or anywhere as long as it's not the first item on a line),
 // C++20 allows "module" to be used. This is provided for backward compatibility, and for
 // simplicity, if someone wants to use py::module for example, that is perfectly safe.
 using module = module_;
 
-#if PY_MAJOR_VERSION >= 3
 PYBIND11_NAMESPACE_BEGIN(detail)
+#if PY_MAJOR_VERSION >= 3
 inline module_ create_top_level_module(const char *name, const char *doc, PyModuleDef *def) {
-    return module_(name, doc, def);
+    def = new (def) PyModuleDef {  // Placement new (not an allocation).
+        /* m_base */     PyModuleDef_HEAD_INIT,
+        /* m_name */     name,
+        /* m_doc */      options::show_user_defined_docstrings() ? doc : nullptr,
+        /* m_size */     -1,
+        /* m_methods */  nullptr,
+        /* m_slots */    nullptr,
+        /* m_traverse */ nullptr,
+        /* m_clear */    nullptr,
+        /* m_free */     nullptr
+    };
+    auto m = PyModule_Create(def);
+    if (m == nullptr)
+        pybind11_fail("Internal error in detail::create_top_level_module()");
```

Comment:
Idea 1: does `throw error_already_set()` work?
Idea 2: somehow include `name` in the error message, to be helpful?

Reply:
Answered in the main thread, but that was maybe not a great idea:

> I'm now just looking at the error message, to finish this PR, but I can't find if a nullptr mean the error indicator is already set?
> https://docs.python.org/3/c-api/module.html#c.PyModule_Create
>
> EDIT: If it definitely does, we should throw py::error_already_set to be fully correct!

Reply:
Looking at the 3.7 sources, I see some code paths set an error for sure, but it's difficult to know if that's always the case. The bomb-proof solution would be something like `if (PyErr_Occurred()) throw error_already_set();` and `pybind11_fail()` as a fallback, with `name` in the error message.
I kind of feel it's worth it: in case this fires it's most likely a difficult to debug situation,  and showing the original error (root cause) or the module `name` could save a lot of time troubleshooting.
Unless we can convince the Python developers to document that a NULL return guarantees the error is set ...

Reply:
You might not need to add name to the error message, if it usually or hopefully always works with `error_already_set()`, as that makes `pybind11_fail()` a last-ditch fallback.

Reply:
(Assuming error_already_set would show the name?)

Reply:
Myeah, or we just trust that this part of the API works in the same way? An `error_already_set` won't cause things to be messed up too badly.

I can't both throw `error_already_set` ánd have the `name` in there, though. So that's only as a fall-back?

Reply:
> (Assuming error_already_set would show the name?)

I don't know; we don't control that, since that the message is already set by Python. It might be something as cryptic as "didn't manage to create a `str` from this `const char*`"

Reply:
OK, didn't add the name, as the `pybind11_fail` is only our fallback option. We can add it, if we find out that (and when!) the control flow can reach that statement?